### ISSUE TITLE
Rerender performance took a hit since version 6.8.4

### ISSUE THREAD

**Issue by josias-r** [react-hook-form/react-hook-form#3379]
**Describe the bug**
We recently came across (unrelated) performance issues in our application where inputs were lagging. I looked into those a fixed some stuff from our end. But then I noticed that after I upgraded react-hook-form to the latest version (6.10.1) some inputs were lagging again.

**To Reproduce**
Our forms are set-up using your Controller components + setValue method for event handlers. We don't use native HTML inputs.

**Expected behavior**
A version upgrade should not increase the amount of rerenders without good reason.

**Screenshots**
43 rerenders, no input lag - version 6.8.4
<img width="1720" alt="6 8 6" src="https://user-images.githubusercontent.com/11424820/98374911-a9488f80-2041-11eb-8b66-29c4a7802164.png">

129 rerenders, input lag - version 6.8.6
<img width="1720" alt="6 8 4" src="https://user-images.githubusercontent.com/11424820/98374846-92a23880-2041-11eb-875c-2b72360fd0b1.png">

(I don't know about 6.8.5 because it throws me some type errors, maybe it got introduced in that version already, but any version later then 6.8.6 definitely has this problem)

This is a problem because there are noticeable input lags for our rather large forms 😄 

PS. ignore the nested Controller Elements, I know that's not really smart to do, but it is necessary for a few forms in our application for some complicated reason 😛

**Comment by bluebill1049** [comment]
can you share a reproducible CSB? It would help with this issue.

**Comment by bluebill1049** [comment]
btw, we do have perf test in the lib but nothing pick up, so i think i need to see the implementation to understand the root cause.

**Comment by josias-r** [comment]
@bluebill1049 I'm still working on dumbing down our production example to a reproduction example. But what I found out so far, that it has specifically something to do with deconstruction the `isValid` boolean returned from the hook.

When I comment out `formState: { isValid },` our form suddenly has a lot less renders. For some reason, the root form component re-renders when this isValid boolean is deconstructed.

I even dumbed it down so that this boolean isn't even used anymore, and the problem persists. You have any idea what could be the cause? I'll try to get a code sandbox example still, but it's gonna take me some time 😛

**Comment by bluebill1049** [comment]
thanks, @josias-r are you using schema or build in validators?

**Comment by josias-r** [comment]
In my dumbed-down example, there are no validation rules whatsoever. It's just this hook call:
```
const {
  setValue,
  formState: { isValid },
  control,
} = useForm<FormShape>({
  defaultValues: {},
});
```
+ this controller:
```
<Controller
  name="aField"
  control={control}
  render={({ value: postalCodeVal }) => {
    return (
      <OurComplexComponent
        onSubvaluesChange={fields => {
          setValue("aField", fields.a_field || undefined, {
            shouldDirty: true,
            shouldValidate: true,
          });
        }}
      />
    );
  }}
/>
```

A single controller wrapping a complex form element that calls the setValue for an input field inside it.
Somehow when I use HTML input elements instead of our complex form element, the problem doesn't exist that the root component rerenders.

**Comment by josias-r** [comment]
So the only two factors are this formState.isValid deconstructing and our complex component that cause the root component to re-render in newer versions of react-hook-form

**Comment by amankkg** [comment]
Not sure if it's related but I faced another performance issue. Unmounting a component triggers form/useForm re-render. Here is repro https://codesandbox.io/s/inspiring-christian-24gg8

In our production, we do not face such issues, so it is only in development and probably somehow related to hot-reload.

**Comment by bluebill1049** [comment]
@josias-r so you are using in build validation.

I think this is suspicious::

```
setValue("aField", fields.a_field || undefined, {
  shouldDirty: true,
  shouldValidate: true,
});
```

**Comment by josias-r** [comment]
@bluebill1049 How exactly do you differ "build" and "schema" validation?
Your comment made me think though, and I found out that because I update multiple values at once in that "onSubvaluesChange" prop, the "shouldValidate" causes a full form rerender for each of the setValues. I was able to get rid of the input lag by calling "trigger" once after the setting the values, which reduces the amounts of full rerenders and gets rid of the input lag.

What I still don't understand is why it does fully rerender. It should only re-render if the isValid state changes, right?

**Comment by bluebill1049** [comment]
@josias-r it shouldn't only trigger re-render when `isValid` changed.

**Comment by josias-r** [comment]
The current behaviour is that it fully rerenders the main form component, even if the validators don't respond with a different error message (or in my case no error messages at all). The validations should run every time of course and if the validation results change it should re-render, I see no reason to re-render otherwise. But if there is a good reason for that, I think we can close this 😁

What is still weird to me that I can only see this behaviour with our Complex component. I see no full re-renders with something more simple like an input + the setValue.

Anyways, the single "trigger" call to validate resolved my issue with a visible input lag, so I don't want to waste ur time further, thanks for helping out! 😁

**Comment by bluebill1049** [comment]
any chance to share that complex component in a codesandbox? Happy to investigate the reason behind, it shouldn't' matter complex or single input, they run through the same logic.

**Comment by josias-r** [comment]
I will check later today if I can share it. It's a pretty large and ugly old class component, which we want to refactor at some point anyway. I need to check how much breaks if I separate that component and how much dependencies it has

**Comment by bluebill1049** [comment]
no worries. I am closing this issue but will follow up once you have an example that i can check out.

**Comment by josias-r** [comment]
Unfortunately, It's not really possible in without an unreasonable amount of work to get that component working in a code sandbox, let alone dumb it down enough to make it possible to debug it.

I'll play around some more in our codebase and if I find some more hints, Ill post a comment.
Thanks for your help and this awesome package!

**Comment by vanyadymousky** [comment]
@amankkg Hey! Also experiencing re-render on every input unmount, it does hurt really bad. Any options you consider or separate issue you've opened? Have you tried v7?

**Comment by amankkg** [comment]
@vanyadymousky Hi, haven't tried v7 yet, we decided not to unmount any components, just set them hidden + disabled

**Comment by vanyadymousky** [comment]
@amankkg Thanks for your answer. For me shouldUnregister: false works yet
